### ISSUE TITLE
[otbn] Add support for pseudo-instructions in the assembler

### ISSUE THREAD

**Issue by rswarbrick** [lowRISC/opentitan#2967]
To do so, we should add some more instructions into `insns.yml`. These don't get an encoding but instead have a rule for generating a sequence of real instructions.

Currently on the list:

  - [ ] `nop`
  - [ ] `li`: This can be expressed as LUI; ADDI, but is a bit difficult because ADDI takes a signed immediate. So "constant materialisation" needs to get there by subtraction in some cases. Not sure how to do that in the YAML: maybe we have a "there's some Python code for this!" escape hatch?

**Comment by felixmiller** [comment]
`ret` would be helpful for `JALR x0, x1, 0`

Another one that would be really helpful (but probably is not that straight forward) is an `endloop` statement. Otherwise the loop instruction has to be adjusted every time an instruction is added or removed to the loop body. Forgetting this (or making errors when counting instructions) creates nasty errors. In the dcrypto assembly[ this is achieved by using parenthesis](https://chromium.googlesource.com/chromiumos/platform/ec/+/refs/heads/cr50_stab/chip/g/dcrypto/dcrypto_bn.c#33). Personally, I would like `endloop` or similar better, but this is matter of taste I guess.

**Comment by rswarbrick** [comment]
I agree about ret.

For "endloop", I was wondering about doing it with a pseudo instruction like this:
```
   loopi_until 10, end_of_loop
   addi       x1, x2, x3
end_of_loop:
```

The idea being that we could emit an instruction count of the form `(end_of_loop - .) >> 2`. I'm not sure whether the binutils assembler can sort out that sort of immediate, though: probably need to try it out.

**Comment by felixmiller** [comment]
Yes, your proposal is even better. This is much better readable with nested loops then.

**Comment by rswarbrick** [comment]
Updated to tick off the operations that are done. Still to do: a handy syntax for loop/loopi.

**Comment by rswarbrick** [comment]
Note to self, after spending a while trying to do the loop/loopi version: It's not as easy as you might hope. The problem is that neither `loop` nor `loopi` fits into an existing RISC-V encoding, so we can't emit either with an `.insn` line. As such, we have to assemble all the way on our own... which means we need to know the position of the target.

This isn't massively difficult, but it will force us to run `otbn-as` in two passes. The first parses the syntax of the file and spots all the labels. The second walks the resulting "AST" and sorts out the instructions. No biggie, but maybe a day's work.
